---
title: "A minimal gene set characterizes TIL specific to diverse tumor antigens across different cancer types"
output: html_document

date: "2024-11-05"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Contents
***
### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[1. Overview](#1.-overview)
### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[2. Input data preparation](#2.-input-data-preparation)
### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[3. Predict MANAscore](#3.-predict-manascore)
### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[4. Define MANAscore^hi^ TIL](#4.-define-manascorehi-til)
### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[5. Define pTRC](#5.-define-ptrc)


## 1. Overview{#1.-overview}
***
Pinpointing tumor-specific T cell clones responsible for immunotherapy responses remains challenging. Based on previous findings that validated mutation-associated neoantigen (MANA)-specific CD8+ tumor-infiltrating lymphocytes (TIL) from several cancer types express high levels of CXCL13 and CD39 (ENTPD1), and low IL7R levels, we develop a ‘MANAscore’ using weighted expression levels of these three genes from lung cancer and melanoma single-cell RNAseq, aiming at prospectively identify MANA-specific TIL. Our three-gene MANAscore algorithm outperforms other RNAseq-based algorithms in identifying validated neoantigen-specific CD8+ clones and is further demonstrated to accurately identify TIL recognizing other classes of tumor antigens, including cancer-testis antigens, endogenous retroviruses, and viral oncogenes, and most of these TIL expressed a tissue resident memory program.

## 2. Input data preparation{#2.-input-data-preparation}
***
Load the data
```R
ser <- readRDS('/the/folder/of/object.rds')
DefaultAssay(ser) <- 'RNA'
```
Run saver to get imputation data, then the quantile normalize and scale the data for each sample, then extract three genes (CXCL13, ENTPD1 and ENTPD1) for each sample
```R
d <- ser[[]] %>% group_by(sample) %>% summarise(n=n()) %>% arrange(n)
print(d)

ser_saver = NULL
dat = NULL
for(i in 1:dim(d)[1]){
  s = d[i,]$sample
  print(s)
  subser = subset(ser, subset=sample==s)
  subser_count = subser[['RNA']]$counts
  dim(subser_count)
  subser_saver = saver(subser_count, ncores = 96,estimates.only = T)
  ser_saver= cbind(ser_saver, subser_saver)
  
  ser_saver_norm <- normalize.quantiles(subser_saver, copy=TRUE)
  colnames(ser_saver_norm) <- colnames(subser_saver)
  rownames(ser_saver_norm) <- rownames(subser_saver)
  
  ser_saver_norm_scale <- scale(ser_saver_norm)
  
  genes <- c('CXCL13','ENTPD1','IL7R')
  
  subdat <- ser_saver_norm_scale[genes,] %>% t() %>% as.data.frame()
  dat <- rbind(dat, subdat)
  
}

write.csv(dat, "dat_3gene_imputation.csv", quote=FALSE)
```
Extract 3 gene non-imputation data from seurat object
```R
RNA_3gene <- ser@assays$RNA$data[genes,] %>% as.matrix() %>% t() %>% as.data.frame()
write.csv(RNA_3gene, "dat_3gene_RNA.csv", quote=FALSE)

```
## 3. Predict MANAscore{#3.-predict-manascore}
***
Option1: Loading the saved models for prediction
```shell
manascore predict input_imputation_3_gene.csv input_non_imputation_3_gene.csv output_imputation_score.csv output_non_imputation_score.csv
```

Option2: Starting with loading ground truth to build the voting models then predict the MANAscore, the models will be saved under current folder
```shell
manascore train_predict input_imputation_3_gene.csv input_non_imputation_3_gene.csv output_imputation_score.csv output_non_imputation_score.csv
```

## 4. Define MANAscore^hi^ TIL{#4.-define-manascorehi-til}
***
We set patient specific cutoffs for imputation MANAscore (MANAscore_i) and non-imputation MANAscore (MANAscore_ni) in tumor tissue because of patient heterogeneity.
```R
m <- ser[[]] %>% group_by(sample) %>% summarise(n=n())
D <- NULL
Cutoff <- NULL
for(z in 1:length(m[,1])){
  print(z)
  s = m[z,]$sample
  cutoff= NULL
  d = ser[[]] %>% filter(sample==s)
  di <- density(d$score_i)
  trough1 <- NULL
  for (i in 2:(length(di$y)-1)) {
    if (di$y[i-1] >= di$y[i] & di$y[i] <= di$y[i+1]) {
      trough1 <- cbind(trough1, c(di$x[i], di$y[i]))
    }
  }
  cutoff1 =  trough1[,dim(trough1)[2]][1]
  d$it = ifelse(d$score_i>cutoff1,'hi','lo')
  dni <- density(d$score_ni)
  trough2 <- NULL
  for (i in 2:(length(dni$y)-1)) {
    if (dni$y[i-1] >= dni$y[i] & dni$y[i] <= dni$y[i+1]) {
      trough2 <- cbind(trough2, c(dni$x[i], dni$y[i]))
    }
  }
  cutoff2 =  trough2[,dim(trough2)[2]][1]
  d$nt = ifelse(d$score_ni>cutoff2,'hi','lo')
  D <- rbind(D,d)
  cutoff = data.frame(sample = s,
                      cutoff_i = cutoff1,
                      cutoff_ni = cutoff2)
  Cutoff = rbind(Cutoff,cutoff)
  ##
  plot1 <- ggscatter(d, x = "score_i", y = "score_ni",size = 0.5,title = ps,
                     add = "reg.line",  # Add regressin line
                     add.params = list(color = "#008B45FF", fill = "lightgray"),
                     conf.int = TRUE # Add confidence interval
  )+
    geom_hline(yintercept=cutoff2,color='red',linetype="dashed",linewidth=1.5)+
    geom_vline(xintercept=cutoff1,color='red',linetype="dashed",linewidth=1.5)+
    scale_y_continuous(expand = c(0,0),limits = c(0,1))+
    scale_x_continuous(expand = c(0,0),limits = c(0,1))+
    # geom_rug()+
    stat_cor(method = "pearson",label.x = 0.5, label.y = 0.05,color="#008B45FF",size=4) +
    labs(x='MANAscore_i',y='MANAscore_ni') +
    theme(axis.title = element_text(size=14,face='bold'),
          axis.text = element_text(size=12),
          plot.title = element_blank())
  dens1 <- ggplot(d, aes(x = score_i)) +
    geom_density() +
    theme_classic()+
    geom_point(aes(x=trough1[,dim(trough1)[2]][1],
                   y=trough1[,dim(trough1)[2]][2]),colour="red",size=3)+
    labs(title=ps)+
    scale_x_continuous(expand=c(0,0),limits = c(0,1))+
    theme(legend.position = "none",axis.title = element_blank(),
          axis.text = element_blank(),
          axis.ticks = element_blank(),
          plot.title = element_text(size=16,hjust=0.5,face='bold'))
  dens2 <- ggplot(d, aes(x = score_ni)) +
    geom_density(alpha = 0.4) +
    theme_classic() +
    geom_point(aes(x=trough2[,dim(trough2)[2]][1],
                   y=trough2[,dim(trough2)[2]][2]),colour="red",size=3)+
    scale_x_continuous(expand=c(0,0),limits = c(0,1))+
    theme(legend.position = "none",axis.title = element_blank(),
          axis.text = element_blank(),
          axis.ticks = element_blank())+
    coord_flip()
  dens2
  dens1 + plot_spacer() + plot1 + dens2 +
    plot_layout(ncol = 2, nrow = 2, widths = c(4, 1), heights = c(1, 4))
  ggsave(paste0('./correlation/s,'_correlation.jpeg'),height=6,width=6)
}

D %>% filter(it == 'hi' & nt =='hi') %>% dim() 
saveRDS(D, 'MANAscore.rds')
```
## 5. Define pTRC{#5.-define-ptrc}
***
Then we further identified the pTRC (potential tumor reactive T cell) as the cells within clone which are with at least 5 MANAscore^hi^ TIL for lung cancer data.


```{=html}
<script>
// Function to add copy buttons to each code block
document.addEventListener("DOMContentLoaded", function() {
  document.querySelectorAll("pre code").forEach(function(block) {
    // Create a copy button and append it to each code block's parent
    const button = document.createElement("button");
    button.className = "copy-button";
    button.innerText = "Copy";
    block.parentNode.style.position = "relative"; // Ensure the button is positioned correctly
    block.parentNode.appendChild(button);

    // Add click event to the button to copy code text
    button.addEventListener("click", function() {
      navigator.clipboard.writeText(block.innerText).then(() => {
        button.innerText = "Copied!";
        setTimeout(() => (button.innerText = "Copy"), 2000);
      }).catch(() => alert("Failed to copy text."));
    });
  });
});
</script>

<style>
.copy-button {
  position: absolute;
  top: 8px;
  right: 8px;
  z-index: 1;
  padding: 4px 8px;
  font-size: 0.9em;
  cursor: pointer;
  background-color: #6c757d; /* Gray background color */
  color: white; /* White text color */
  border: none;
  border-radius: 3px;
}
.copy-button:hover {
  background-color: #5a6268; /* Darker gray on hover */
}
</style>
```